### PR TITLE
Add HDF5 (pt. 2) - merge when passing

### PR DESCRIPTION
**Pull request by jakirkham**
This continues off @groutr's excellent work in this PR ( https://github.com/conda-forge/staged-recipes/pull/192 ) to get a build of HDF5 going.

Initially thought Travis CI was failing builds due to a lack of output from the tests. However, it has been determine that one of the test portions was causing a hang. Namely the `cache` tests. Here we simply patch the `test/Makefile.in` to skip the `cache` test program. This appears to be enough to get Travis CI to run through. We go ahead and apply this patch everywhere even though Travis CI was the only one that needed. My best guess is this test used too much memory, which led to hang.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @ocefpaf 

Maybe it won't be that hacky at all. It turns out there is a `--disable-silent-rules` argument. I just passed this to `./configure`. Hopefully, that won't be too noisy though or we will have different problems with Travis and or the other ones.

File: recipes/hdf5/build.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+./configure --prefix=$PREFIX \
+    --enable-linux-lfs --with-zlib \
+    --with-pthread=yes  --enable-cxx --with-default-plugindir=$PREFIX/lib/hdf5/plugin \
+    --disable-silent-rules  # To make Travis happy with the level of activity.
```
Comment (jakirkham) [inline recipes/hdf5/build.sh]: Here's the argument to add a bit more noise to the Travis CI build. Hopefully, that is the right amount without going overboard the other way.

**Comment by jakirkham** [comment]
Appears this is not an issue with the amount of noise, but actually may be a memory issue with the `cache` test. Attempting to patch this out of the test running step so that maybe we can complete the test suite.

File: recipes/hdf5/build.sh [hunk 2]
```diff
@@ -2,7 +2,8 @@
 
 ./configure --prefix=$PREFIX \
     --enable-linux-lfs --with-zlib \
-    --with-pthread=yes  --enable-cxx --with-default-plugindir=$PREFIX/lib/hdf5/plugin
+    --with-pthread=yes  --enable-cxx --with-default-plugindir=$PREFIX/lib/hdf5/plugin \
+    --disable-silent-rules  # To make Travis happy with the level of activity.
```
Comment (jakirkham) [inline recipes/hdf5/build.sh]: This attempts to increase the noise level.

File: recipes/hdf5/build.sh [hunk 3]
```diff
@@ -2,8 +2,7 @@
 
 ./configure --prefix=$PREFIX \
     --enable-linux-lfs --with-zlib \
-    --with-pthread=yes  --enable-cxx --with-default-plugindir=$PREFIX/lib/hdf5/plugin \
-    --disable-silent-rules  # To make Travis happy with the level of activity.
+    --with-pthread=yes  --enable-cxx --with-default-plugindir=$PREFIX/lib/hdf5/plugin
```
Comment (jakirkham) [inline recipes/hdf5/build.sh]: However, the build still ended up timing out for ~10mins. So, there is something else going on. As it was the `cache` test, expect this to be memory related. Perhaps, it is going to swap, which would explain why this bogs down.

File: recipes/hdf5/test_Makefile.in.patch [hunk 4]
```diff
@@ -0,0 +1,13 @@
+diff --git test/Makefile.in test/Makefile.in
+index 02ab570..f2b9889 100644
+--- test/Makefile.in
++++ test/Makefile.in
+@@ -1071,7 +1071,7 @@ check_SCRIPTS = $(TEST_SCRIPT)
+ # These tests (fheap, btree2) are under development and are not used by
+ # the library yet. Move them to the end so that their failure do not block
+ # other current library code tests.
+-TEST_PROG = testhdf5 lheap ohdr stab gheap cache cache_api \
++TEST_PROG = testhdf5 lheap ohdr stab gheap cache_api \
```
Comment (jakirkham) [inline recipes/hdf5/test_Makefile.in.patch]: As the `cache` test is the one that it bogged down on, simply drop this from the list of test programs to run.

**Comment by jakirkham** [comment]
So, I have added a patch here, which simply does not run the cache program from `make check`. The alternative would be not to run `make check`, which was deemed unacceptable. This at least allows the rest of the tests (other than `cache`) to run. None of those have been seen to have problems on any platform.

This appears to allow for the Mac Travis CI build to complete. Without it, the build hangs and is terminated after no response for ~10mins.

It is unlikely this patch would be accepted upstream as it skips one of their tests. So, I don't plan to submit it. However, a discussion could begin about making this test less resource intensive (assumed to be memory intensive). Depending on the nature of the test, this may or may not be feasible.

**Comment by jakirkham** [comment]
:tada: Travis CI passes! :tada:

File: recipes/hdf5/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+package:
+  name: hdf5
+  version: {{ version }}
+
+source:
+  fn: hdf5-{{version}}.tar.gz
+  url: http://www.hdfgroup.org/ftp/HDF5/current/src/hdf5-{{version}}.tar.gz
+  md5: b8ed9a36ae142317f88b0c7ef4b9c618
+  patches:
+    - test_Makefile.in.patch
```
Comment (pelson) [inline recipes/hdf5/meta.yaml]: I'd want a really good comment in here to explain the rationale. Maybe provide a link back to the original PR too.

P.S. I'm happy for this to be done in the feedstock once merged.

Comment (jakirkham) [inline recipes/hdf5/meta.yaml]: No problem. Are you ok with the rationale?

Comment (pelson) [inline recipes/hdf5/meta.yaml]: Yep. Slightly concerned - maybe we should have an environment variable which allows us to turn stuff on if we are ```conda-forge``` building stuff. Like:

```
patches:
    - test_Makefile.in.patch  # [os.environ.get(conda-forge)]
```

**Pull request merged**